bridgerelayer/chain/substrate/chain.go

```diff
@@ -0,0 +1,102 @@
+package substrate
+
+import (
+	"context"
+	"fmt"
+
+	"golang.org/x/sync/errgroup"
+
+	"github.com/sirupsen/logrus"
+	"github.com/snowfork/polkadot-ethereum/bridgerelayer/chain"
+	"github.com/snowfork/polkadot-ethereum/bridgerelayer/crypto/sr25519"
+	"github.com/spf13/viper"
+)
+
+// Chain ...
+type Chain struct {
+	listener *Listener
+	writer   *Writer
+	conn     *Connection
+}
+
+const Name = "Substrate"
+
+// NewChain ...
+func NewChain(ethMessages chan chain.Message, subMessages chan chain.Message) (*Chain, error) {
+
+	log := logrus.WithField("chain", Name)
+
+	// Validate and load configuration
+	keys := []string{
+		"substrate.endpoint",
+		"substrate.private-key",
+		"substrate.block-retry-limit",
+		"substrate.block-retry-interval",
+	}
+	for _, key := range keys {
+		if !viper.IsSet(key) {
+			return nil, fmt.Errorf("Config key %q not set", key)
+		}
+	}
+	endpoint := viper.GetString("substrate.endpoint")
+	secret := viper.GetString("substrate.private-key")
+	blockRetryLimit := viper.GetUint("substrate.block-retry-limit")
+	blockRetryInterval := viper.GetUint("substrate.block-retry-interval")
+
+	// Generate keypair from secret
+	kp, err := sr25519.NewKeypairFromSeed(secret, "")
+	if err != nil {
+		return nil, err
+	}
+
+	conn := NewConnection(endpoint, kp.AsKeyringPair(), log)
+
+	listener := NewListener(
```

Comment:
```suggestion
	listener := NewListener(conn, subMessages, blockRetyLimit, blockRetryInterval, log)
```
Nit: any reason this is broken up onto multiple lines? The line length seems manageable.